src/content/1.7/modules/concepts/controllers/admin-controllers/override-decorate-controller.md

```diff
@@ -0,0 +1,171 @@
+{{% /notice %}}
+
+## Decorate the controller
+
+It is better to use [decoration](https://symfony.com/doc/3.4/service_container/service_decoration.html) which keeps the behavior of the decorated controller while allowing you to extend its behavior by modifying received inputs and outputs.
+
+Back to the controller responsible for the page "Improve > Design > CMS Pages" which is registered with service ID `PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController`.
+
+With the following configuration item, we can decorate it with a custom controller in our module:
+```yaml
+# modules/your-module/config/services.yml
+  custom_controller:
+    class: MyModule\Controller\DemoController
+    decorates: PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController
+    arguments: ['@custom_controller.inner']
+```
+
+Thanks to this, whenever Symfony forwards a request to the Core controller `PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController` it will be forwarded to `DemoController` instead. But what is different with overriding is that the decorated CmsPageController is injected into the constructor of DemoController, and we can use it:
+```php
+<?php
+
+namespace MyModule\Controller;
+
+use PrestaShop\PrestaShop\Core\Search\Filters\CmsPageCategoryFilters;
+use PrestaShop\PrestaShop\Core\Search\Filters\CmsPageFilters;
+use PrestaShopBundle\Controller\Admin\FrameworkBundleAdminController;
+use PrestaShopBundle\Controller\Admin\Improve\Design\CmsPageController;
+use Symfony\Component\HttpFoundation\Request;
+
+class DemoController extends FrameworkBundleAdminController
+{
+    /**
+     * @var CmsPageController
+     */
+    private $decoratedController;
+
+    public function __construct(CmsPageController $decoratedController)
+    {
+        $this->decoratedController = $decoratedController;
+    }
+
+    public function indexAction(CmsPageCategoryFilters $categoryFilters, CmsPageFilters $cmsFilters, Request $request)
+    {
+        return $this->decoratedController->indexAction($categoryFilters, $cmsFilters, $request);
+    }
+}
+```
+In this example I do nothing else than returning the exact output from the decorated controller.
+However I could modify the input request or I could modify the output given by decorated controller before returning it. Example:
+```php
+    public function indexAction(CmsPageCategoryFilters $categoryFilters, CmsPageFilters $cmsFilters, Request $request)
+    {
+        $output = $this->decoratedController->indexAction($categoryFilters, $cmsFilters, $request);
+
+        $myService = $this->getMyPaymentService();
+        $output = $this->injectMyData($myService, $output);
+
+        return $output;
+    }
+```
+
+
+{{% notice tip %}}
+if you have trouble writing the right service configuration for your decoration, you can use Symfony debugger to dump the routes with `php bin/console debug:container`. It can also be helpful to find the service ID of the controller.
```

Comment:
```suggestion
If you have trouble writing the right service configuration for your decoration, you can use Symfony debugger to dump the routes with `php bin/console debug:container`. It can also be helpful to find the service ID of the controller.
```